### ISSUE TITLE
Approved videos disappear.

### ISSUE THREAD

**Issue by psyciknz** [photoprism/photoprism#618]
So I went though some of my review items, and looked at the videos, gave a couple of them some manual categories....wondered why I couldn't set a country (why I was browsing the issues).

But now I go back and look at the video library, and it's empty, and wants me to approve the videos again.

How can I figure out what's happening here.  

I did find this thought: November 17, 2020, 12:31:43 PM GMT+13
photo: open /photoprism/storage/sidecar/iPhone David/2019/2019-12 December iPhone/IMG_5334.yml: permission denied (update yaml)

**Comment by lastzero** [comment]
Looks like a file permission issue?

**Comment by psyciknz** [comment]
yeah, so lack of writing the yaml file would mean the videos go back into the review pile?  I guess that's more my question.

**Comment by lastzero** [comment]
If we don't handle this specific type of error yet, that would be theoretically possible. You also need write permissions for most other operations like indexing. Make sure to test with today's Docker image.

**Comment by psyciknz** [comment]
Yeah I think mine updated this morning.

But I went throudh and did a chwon to my account (the one used for uid/pgid).  But now running a photoprisim index command in docker, and I'm getting:
```
panic: Error 1018: Can't read dir of './photoprism/' (errno: 13 "Permission denied")

goroutine 1 [running]:
github.com/jinzhu/gorm.mysql.HasTable(0x125dc00, 0xc0135b0820, 0xc0175643e0, 0xc0175643e0, 0x5, 0x7)
        /go/pkg/mod/github.com/jinzhu/gorm@v1.9.16/dialect_mysql.go:180 +0x372
github.com/jinzhu/gorm.(*Scope).autoMigrate(0xc013502180, 0x1015060)
        /go/pkg/mod/github.com/jinzhu/gorm@v1.9.16/scope.go:1268 +0xbb
github.com/jinzhu/gorm.(*DB).AutoMigrate(0xc0135121a0, 0xc01349f6b0, 0x1, 0x1, 0xc01349f738)
        /go/pkg/mod/github.com/jinzhu/gorm@v1.9.16/main.go:689 +0x96
github.com/photoprism/photoprism/internal/entity.Types.Migrate(0xc01344bbf0)
        /go/src/github.com/photoprism/photoprism/internal/entity/entity.go:100 +0xf6
github.com/photoprism/photoprism/internal/entity.MigrateDb()
        /go/src/github.com/photoprism/photoprism/internal/entity/entity.go:134 +0x2d
github.com/photoprism/photoprism/internal/config.(*Config).InitDb(0xc0004b7600)
        /go/src/github.com/photoprism/photoprism/internal/config/db.go:110 +0x49
github.com/photoprism/photoprism/internal/commands.indexAction(0xc0004c02c0, 0x0, 0x0)
        /go/src/github.com/photoprism/photoprism/internal/commands/index.go:45 +0x130
github.com/urfave/cli.HandleAction(0xf4d720, 0x1120478, 0xc0004c02c0, 0xc0004c02c0, 0x0)
        /go/pkg/mod/github.com/urfave/cli@v1.22.5/app.go:524 +0xfd
github.com/urfave/cli.Command.Run(0x10a1676, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x10d482e, 0x27, 0x0, ...)
        /go/pkg/mod/github.com/urfave/cli@v1.22.5/command.go:173 +0x58e
github.com/urfave/cli.(*App).Run(0xc0004ca000, 0xc0000a4060, 0x3, 0x3, 0x0, 0x0)
        /go/pkg/mod/github.com/urfave/cli@v1.22.5/app.go:277 +0x7e8
main.main()
        /go/src/github.com/photoprism/photoprism/cmd/photoprism/photoprism.go:71 +0x529
```

Unsure  if it's the image, or me running chown.

**Comment by psyciknz** [comment]
I had a look around the dirs and created some files and couldn't actually see where it was getting an issue.  I tried the sidecars dir, originals, db and default ./
```
root@645fad8c059d:/photoprism# touch temp
root@645fad8c059d:/photoprism# ls -la
total 36
drwxrwxrwx 1 photoprism photoprism 4096 Nov 17 00:02 .
drwxr-xr-x 1 root       root       4096 Nov 16 23:24 ..
-rw-rw-rw- 1 photoprism photoprism  220 Feb 25  2020 .bash_logout
-rw-rw-rw- 1 photoprism photoprism 3771 Feb 25  2020 .bashrc
-rw-rw-rw- 1 photoprism photoprism  807 Feb 25  2020 .profile
drwxrwxrwx 1 photoprism photoprism 4096 Nov 16 18:56 assets
drwxrwxrwx 1 photoprism photoprism 4096 Nov 16 18:56 bin
drwxrwxrwx 2 photoprism photoprism 4096 Nov 16 18:56 import
drwxr-xr-x 2 photoprism root          0 Nov 16 23:58 originals
drwxr-xr-x 7 photoprism photoprism 4096 Nov 16 23:58 storage
-rw-r--r-- 1 root       root          0 Nov 17 00:02 temp
```

**Comment by psyciknz** [comment]
So after updating those permissions, they seem to have reverted, as now the file has gone from videos again.
photo: open /photoprism/storage/sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.yml: permission denied (update yaml)
It's created the json file though when it updated.  But theres no yml.  Also I notice that sidecar files are created by root.  Any ideas what it shold be set to?

**Comment by psyciknz** [comment]
So I managed to get photoprism to update a yml file...and it's created by photoprism.  So what do I set the permissions to on the host system to allow this?

**Comment by psyciknz** [comment]
So now about 0 mins later, I've seen this in teh logs:
time="2020-11-18T01:47:19Z" level=debug msg="meta-worker: starting routine check"
time="2020-11-18T01:47:19Z" level=debug msg="moments: index contains 289 photos and 0 videos, using threshold 9"

And when I look at videos, teh 13 videos I'd approved have now gone.   I really dont' understand what's happening.

**Comment by psyciknz** [comment]
So i've been watching it, and playing in the db.  I got the query of fields from 
```
Select("SUM(photo_type = 'video' AND photo_quality >= 0 AND photo_private = 0) AS videos, SUM(photo_type IN ('image','raw','live') AND photo_quality < 3 AND photo_quality >= 0 AND photo_private = 0) AS review, SUM(photo_quality = -1) AS hidden, SUM(photo_type IN ('image','raw','live') AND photo_private = 0 AND photo_quality >= 0) AS photos, SUM(photo_favorite = 1 AND photo_quality >= 0) AS favorites, SUM(photo_private = 1 AND photo_quality >= 0) AS private").
```
After a video is approved, I saw the quality is 4.
```
MariaDB [photoprism]> select photo_uid,id,photo_name ,photo_private,photo_quality from photos where photo_uid = 'pqjx87o184ackmcg';
+------------------+-----+------------+---------------+---------------+
| photo_uid        | id  | photo_name | photo_private | photo_quality |
+------------------+-----+------------+---------------+---------------+
| pqjx87o184ackmcg | 331 | IMG_4938   |             0 |             4 |
+------------------+-----+------------+---------------+---------------+
```

But then after the moments call, the quality has dropped to -1 so it doesn't show up in the videos count.
```
MariaDB [photoprism]> select photo_uid,id,photo_name ,photo_private,photo_quality from photos where photo_uid = 'pqjx87o184ackmcg';
+------------------+-----+------------+---------------+---------------+
| photo_uid        | id  | photo_name | photo_private | photo_quality |
+------------------+-----+------------+---------------+---------------+
| pqjx87o184ackmcg | 331 | IMG_4938   |             0 |            -1 |
+------------------+-----+------------+---------------+---------------+
````

What is causing that quality to be reset back to -1, which means it's not approved any more?

**Comment by lastzero** [comment]
Lots of different issues in here. Trying to isolate them and see what we can do! :+1: 

One thing is for sure: PhotoPrism needs write permissions for the storage folder (and the database), that's not something we can ultimately fix in our code. Just try to work around it, show a more helpful error message, etc

**Comment by lastzero** [comment]
The Docker container can be started as root or a specific users, see our example config files: 

https://dl.photoprism.org/docker/docker-compose.yml
https://dl.photoprism.org/docker/arm64/docker-compose.yml

If you use the environment variables, the container will start as root and then switch the user before starting PhotoPrism. If you set the user directly via Docker, the container starts as this user and can't be changed any more. You can also define a UMASK to set the permissions of newly created files.

Note that the user `photoprism` inside the container isn't necessarily the same as on your host (if it exists).

Videos should NOT show up in Photos > Review at all, see

https://github.com/photoprism/photoprism/blob/develop/internal/entity/photo_quality.go#L64

If the picture type is not image, the minimum quality score is 3. Only pictures with a score less than 3 require a review. The only reason it might be -1 is that indexing didn't work properly due to write permissions issues.

**Comment by lastzero** [comment]
Removing the bug label until it's clear if a fix is needed to solve your issues and if yes, what specifically.

**Comment by psyciknz** [comment]
Hi.  So after the last lot of permission warnings and fixes I do not have those errors any more.  So if there is still a permissions error, it’s not one being reported in the error log (in the ui), and I haven’t seen anything in the stdout log (when in docker logs)

I can try the umask option,  I have no issues with running photoprism reset on the db if you think it will help.

**Comment by lastzero** [comment]
Starting with a fresh database may help. Ideally, PhotoPrism should handle these errors gracefully so that the index doesn't get damaged in any way. Only testing will tell.

**Comment by psyciknz** [comment]
Ok I'll do that.

So can we go over my docker-compose settings for users.  I have
```
UID: 1000
GID: 1000
# UMASK: 00000
```
Set currently, where UID 1000 is my main user account.
The photoprism storage folder is currently a+rwx.
What should that be set to (since a+rwx is a little open, but its fine), and what should I then set the umask to?

**Comment by psyciknz** [comment]
So I've set everything, and it's all clean from a permissions point of view as far as I can tell.

I went in after index and approved the one video I'd been testing with and can see it wrote the yml file here
```
time="2020-11-22T18:51:14Z" level=debug msg="photo: changed title of pqk7o753ie9lik1n to “July iPhone / 2019”"
time="2020-11-22T18:51:14Z" level=info msg="photos: found 1 results for id:pqk7o753ie9lik1n merged:true [1.055948ms]"
time="2020-11-22T18:51:14Z" level=info msg="Changes successfully saved"
time="2020-11-22T18:51:14Z" level=debug msg="photo: updated yaml file IMG_4938.yml"
```
Then I can see the meta worker has run, and I had the query for checking the quality again for that approved video:
This is before an dafter the meta worker ran, with no errors in the error log, or in stdout.
```
MariaDB [photoprism]> select photo_uid,id,photo_name ,photo_private,photo_quality from photos where photo_uid = 'pqk7o753ie9lik1n';
+------------------+-----+------------+---------------+---------------+
| photo_uid        | id  | photo_name | photo_private | photo_quality |
+------------------+-----+------------+---------------+---------------+
| pqk7o753ie9lik1n | 330 | IMG_4938   |             0 |             4 |
+------------------+-----+------------+---------------+---------------+
1 row in set (0.000 sec)

MariaDB [photoprism]> select photo_uid,id,photo_name ,photo_private,photo_quality from photos where photo_uid = 'pqk7o753ie9lik1n';
+------------------+-----+------------+---------------+---------------+
| photo_uid        | id  | photo_name | photo_private | photo_quality |
+------------------+-----+------------+---------------+---------------+
| pqk7o753ie9lik1n | 330 | IMG_4938   |             0 |            -1 |
+------------------+-----+------------+---------------+---------------+
```

**Comment by lastzero** [comment]
Can you check the files table for matching rows? Maybe a JPEG preview could not be created?

**Comment by psyciknz** [comment]
There is a row (well two actually):
```
| id  | photo_id | photo_uid        | instance_id | file_uid         | file_name                                              | file_root | original_name | file_hash                                | file_size | file_codec | file_type | file_mime | file_primary | file_sidecar | file_missing | file_portrait | file_video | file_duration | file_width | file_height | file_orientation | file_projection | file_aspect_ratio | file_main_color | file_colors | file_luminance | file_diff | file_chroma | file_error | mod_time   | created_at          | created_in | updated_at          | updated_in | deleted_at |
+-----+----------+------------------+-------------+------------------+--------------------------------------------------------+-----------+---------------+------------------------------------------+-----------+------------+-----------+-----------+--------------+--------------+--------------+---------------+------------+---------------+------------+-------------+------------------+-----------------+-------------------+-----------------+-------------+----------------+-----------+-------------+------------+------------+---------------------+------------+---------------------+------------+------------+
| 374 |      330 | pqk7o753ie9lik1n |             | fqk7o75s8izqie5v | iPhone David/2019/2019-07 July iPhone/IMG_4938.mp4     | /         |               | bf83c41a3ddda2e8c04d59db9d73fc86b97fb262 |  69111129 | avc1       | mp4       | video/mp4 |            0 |            0 |            0 |             0 |          1 |   39000000000 |       1920 |        1080 |                1 |                 |              1.78 |                 |             |                |         0 |           0 |            | 1603716334 | 2020-11-22 18:47:29 |  234358330 | 2020-11-22 18:47:29 |          0 | NULL       |
| 375 |      330 | pqk7o753ie9lik1n |             | fqk7o75ctmrrctrl | iPhone David/2019/2019-07 July iPhone/IMG_4938.mp4.xmp | /         |               | 8bac06be8b150160d52a44ad66dfad23cc4b5e98 |      1207 |            | xmp       |           |            0 |            1 |            0 |             0 |          0 |             0 |          0 |           0 |                1 |                 |                 0 |                 |             |                |         0 |           0 |            | 1603828551 | 2020-11-22 18:47:29 |  656054517 | 2020-11-22 18:47:29 |          0 | NULL       |
```

Sorry, wasn't quite sure what specific fields you were interested in.  The 2nd row looks like the xmp file.

**Comment by lastzero** [comment]
If a JPEG is missing, quality is set to -1 as no image is available as preview and to analyze the quality, colors, objects,... Basically everything our software does. Can sidecar files be written? Is JPEG conversion turned on?

**Comment by lastzero** [comment]
Maybe ffmpeg can't read the video file?

**Comment by psyciknz** [comment]
Both these exist
```
-rwxr-xr-x 1 david david 7356 Nov 23 07:47 'sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.json'
-rwxrwxr-x 1 david david  340 Nov 23 07:59 'sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.yml'
```
And I can edit them now (added that user parm from the docker-compose.yml, and set the umask to 0002).  So i assume that means all good.  When I approved the video there was no error about updating the yml that I've seen in the past.

Where is jpeg conversion turned on?  I don't see a keyword for that in the compose file (nor the example).

**Comment by lastzero** [comment]
See Settings I our UI. You. Might also see an error or warning in the logs. Maybe turn debug mode on for this.

**Comment by psyciknz** [comment]
Ok, so in the UI, that convert to Jpeg was off.  Turned it on now.
Debug is on (set in docker-compose)  no errors in any logs as far as I can tell.Should I reset the DB, or just approve that video again?

**Comment by lastzero** [comment]
Index again if convert was off.

**Comment by psyciknz** [comment]
So I've reindexed.   I think these are all the log messages pertaining to ath one image:
```
DEBU[2020-11-22T19:35:35Z] convert: iPhone David/2019/2019-07 July iPhone/IMG_4938.mp4 -> IMG_4938.jpg
DEBU[2020-11-22T19:35:35Z] index: IMG_4938.jpg created
INFO[2020-11-22T19:35:35Z] mediafile: 11 thumbnails created for IMG_4938 [214.153798ms]
INFO[2020-11-22T19:35:36Z] index: updated main mp4 file “iPhone David/2019/2019-07 July iPhone/IMG_4938.mp4”
DEBU[2020-11-22T19:35:36Z] metadata: original name IMG_4938.mp4.xmp does not match IMG_4938.mp4 (exiftool)
DEBU[2020-11-22T19:35:36Z] mediafile: exif not supported: IMG_4938.mp4.xmp
INFO[2020-11-22T19:35:36Z] index: updated related xmp file IMG_4938.mp4.xmp
DEBU[2020-11-22T19:35:36Z] metadata: failed parsing IMG_4938.jpg, starting brute-force search (parse jpeg)
DEBU[2020-11-22T19:35:36Z] metadata: original name IMG_4938.jpg does not match IMG_4938.mp4 (exiftool)
DEBU[2020-11-22T19:35:36Z] mediafile: metadata: no exif header in IMG_4938.jpg (search and extract)
DEBU[2020-11-22T19:35:36Z] index: image classification took 230.588286ms
```
and
```
DEBU[2020-11-22T19:35:37Z] index: updated yaml file IMG_4938.yml
DEBU[2020-11-22T19:35:37Z] mediafile: no new thumbnails created for IMG_4938 [88.504µs]
INFO[2020-11-22T19:35:37Z] index: added related jpg file IMG_4938.jpg
```


Ahh I think we have success!!  Now after indexing I actually have videos......they only use to appear in hidden.  Oh and they have a location now, they didn't before - just lat/long co-ords.  So only after the thumbnails are converted will it process/categorise the the video.....

**Comment by psyciknz** [comment]
And I also have a jpg file in the sidecar dir.
```
-rw-r--r-- 1 david david 45029 Nov 23 08:35 'sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.jpg'
-rwxr-xr-x 1 david david  7356 Nov 23 07:47 'sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.json'
-rwxrwxr-x 1 david david   416 Nov 23 08:35 'sidecar/iPhone David/2019/2019-07 July iPhone/IMG_4938.yml'
```

**Comment by lastzero** [comment]
Sure, how else would you implement this? Implement native support for all file formats and keep everything in memory? 😉

**Comment by psyciknz** [comment]
No I think the implementation is fine.  I think that convert option was off by default (I don't remember turning it off).  The only thing that I could suggest to help new users would be a comment around it in the docs, that if you have videos, turn that option on to have them processed.

Thanks for helping me nut it out.  Now I can get on to more interesting things....I'd like to dig into the xmp implementation (would have to quickly learn so go first, and figure out how development works).

**Comment by lastzero** [comment]
Should be on by default, need to verify that.

**Comment by datvance** [comment]
Not sure of the etiquette of replying to closed issues, but I just wanted to say that I'm running into this same issue. I have uid and gid set in docker-compose to my main host system user, and everything in the storage folder is created as that user ... except for sidecar directories and files, which are created as root:root. Not sure why, only possible thing I can think of is indexing in the browser versus indexing via "docker-compose exec photoprism photoprism index". Perhaps running via docker-compose does not use the uid/gid specified in the docker-compose? Just a guess.

**Comment by psyciknz** [comment]
I updated the user in the docker-compose file.  There was another place.
```
services:
  photoprism:
    image: photoprism/photoprism:latest
    restart: unless-stopped
    # Uncomment and edit the following line to set a specific user / group id:
    # user: "1000:1000"
```

Up the top of the sample docker-compose file.  https://dl.photoprism.org/docker/docker-compose.yml

**Comment by datvance** [comment]
@psyciknz Ah, thank you, I did not see that but will try it.

Also, I was mistaken above when I said only the sidecar directory was affected. My cache directory looks like this:

```
dv@cheyenne:~/services/photoprism$ ls -l .photoprism/cache/thumbnails/3
total 64
drwxr-xr-x 17 dv   dv   4096 Nov 30 00:08 0
drwxr-xr-x 16 root root 4096 Nov 29 22:51 1
drwxr-xr-x 18 dv   dv   4096 Nov 29 22:44 2
drwxr-xr-x 16 dv   dv   4096 Nov 29 06:41 3
drwxr-xr-x 15 root root 4096 Nov 29 21:43 4
drwxr-xr-x 15 dv   dv   4096 Nov 29 22:57 5
drwxr-xr-x 13 dv   dv   4096 Nov 29 22:11 6
drwxr-xr-x 16 dv   dv   4096 Nov 29 19:33 7
drwxr-xr-x 17 dv   dv   4096 Nov 29 21:58 8
drwxr-xr-x 14 root root 4096 Nov 30 00:23 9
drwxr-xr-x 17 dv   dv   4096 Nov 29 23:57 a
drwxr-xr-x 15 root root 4096 Nov 29 23:01 b
drwxr-xr-x 14 root root 4096 Nov 29 21:43 c
drwxr-xr-x 17 dv   dv   4096 Nov 29 22:10 d
drwxr-xr-x 16 dv   dv   4096 Nov 29 22:51 e
drwxr-xr-x 16 dv   dv   4096 Nov 29 21:30 f
```

**Comment by lastzero** [comment]
Maybe those were created earlier with different settings? Changing the config won't fix permissions of existing files and folders.
